### ISSUE TITLE
Too long #! lines with bytecode (binary data)

### ISSUE THREAD

**Issue by scemama** [LLNL/spack#16803]
OCaml can create bytecode executables. These are binary files, where the 1st line contains the path to the bytecode interpreter:
```
#!/path/to/ocamlrun
```
Bytecode executables are required for the compilation of the OCaml compiler, so for the installation of the ocaml package.

Because of spack's very long paths, the length of this line can be larger than the system's limit. This is properly handled by Spack for scripts (Python for example), but for OCaml bytecode the data following the 1st ine is binary, and the workaround fails.

This problem of long paths was solved in the latest version of the OCaml compiler, but will always be present with versions of OCaml < 4.10.0.

 
### Steps to reproduce the issue

```console
$ mkdir very_long_path_for_making_spack_crash
$ cd very_long_path_for_making_spack_crash
$ git clone spack
$ spack install ocaml@4.06.0
```

### Error Message

```console
==> 12041: ocaml: Building ocaml [Package]
==> 12041: ocaml: Executing phase: 'install'
==> Error: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xdf in position 160: invalid continuation byte

/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/scemama/spack-stage/spack-stage-ocaml-4.06.0-hmsefzexeljw2eyplfhh3xl7jydyrchd/spack-build-out.txt

==> Error: Failed to install ocaml due to ChildError: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xdf in position 160: invalid continuation byte
/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/scemama/spack-stage/spack-stage-ocaml-4.06.0-hmsefzexeljw2eyplfhh3xl7jydyrchd/spack-build-out.txt
Traceback (most recent call last):
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/installer.py", line 1115, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/hooks/__init__.py", line 59, in __call__
    hook(*args, **kwargs)
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/hooks/sbang.py", line 115, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/hooks/sbang.py", line 102, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/home/scemama/very_long_path_for_making_spack_crash/lib/spack/spack/hooks/sbang.py", line 40, in filter_shebang
    original = original.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xdf in position 160: invalid continuation byte

```

### Information on your system

```console
$ spack debug report
usage: spack debug [-h] SUBCOMMAND ...
spack debug: error: argument SUBCOMMAND: invalid choice: 'report' choose from:
    create-db-tarball

$ git log
commit b472c1380331cc414582f8f5bcb9045e468457c9 (origin/develop)
Author: Sergey Kosukhin <sergey.kosukhin@mpimet.mpg.de>
Date:   Tue Mar 3 22:13:38 2020 +0100

$ python --version
Python 3.7.0

$ uname -a
Linux lpqlx139 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u4 (2019-07-19) x86_64 GNU/Linux

$ cat /etc/issue
Debian GNU/Linux 10 \n \l

$ spack find
==> 2 installed packages
-- linux-debian10-haswell / gcc@8.3.0 ---------------------------
ncurses@6.1  pkgconf@1.6.3

```

### Additional information

When I install spack in my home directory, the path is just small enough to make things work.
This is why I installed spack in a dummy long-named directory.

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by scemama** [comment]
Issue #16772 is related to this bug

**Comment by haampie** [comment]
This is fixed by #26793
